# Problem Statement:
There is techinical details provided in dataset, on that basic, we have to build model to predict car price.
Details include:
1. ID
2. Levy
3. Manufacturer
4. Model
5. Prod. year
6. Category
7. Leather interior
8. Fuel type
9. Engine volume
10. Mileage
11. Cylinders
12. Gear box type
13. Drive wheels
14. Doors
15. Wheel
16. Color
17. Airbags


# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import LabelEncoder,StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score,mean_absolute_error, mean_squared_error

# Imporing Data

This data I have taken from **Kaggle**

In [2]:
data = pd.read_csv("car_price_prediction.csv")
data

ID  Price  Levy   Manufacturer    Model  Prod. year   Category  \
0      45654403  13328  1399          LEXUS   RX 450        2010       Jeep   
1      44731507  16621  1018      CHEVROLET  Equinox        2011       Jeep   
2      45774419   8467     -          HONDA      FIT        2006  Hatchback   
3      45769185   3607   862           FORD   Escape        2011       Jeep   
4      45809263  11726   446          HONDA      FIT        2014  Hatchback   
...         ...    ...   ...            ...      ...         ...        ...   
19232  45798355   8467     -  MERCEDES-BENZ  CLK 200        1999      Coupe   
19233  45778856  15681   831        HYUNDAI   Sonata        2011      Sedan   
19234  45804997  26108   836        HYUNDAI   Tucson        2010       Jeep   
19235  45793526   5331  1288      CHEVROLET  Captiva        2007       Jeep   
19236  45813273    470   753        HYUNDAI   Sonata        2012      Sedan   

      Leather interior Fuel type Engine volume    Mileage  Cylinders  \
0                  Yes    Hybrid           3.5  186005 km        6.0   
1                   No    Petrol             3  192000 km        6.0   
2                   No    Petrol           1.3  200000 km        4.0   
3                  Yes    Hybrid           2.5  168966 km        4.0   
4                  Yes    Petrol           1.3   91901 km        4.0   
...                ...       ...           ...        ...        ...   
19232              Yes       CNG     2.0 Turbo  300000 km        4.0   
19233              Yes    Petrol           2.4  161600 km        4.0   
19234              Yes    Diesel             2  116365 km        4.0   
19235              Yes    Diesel             2   51258 km        4.0   
19236              Yes    Hybrid           2.4  186923 km        4.0   

      Gear box type Drive wheels   Doors             Wheel   Color  Airbags  
0         Automatic          4x4  04-May        Left wheel  Silver       12  
1         Tiptronic          4x4  04-May        Left wheel   Black        8  
2          Variator        Front  04-May  Right-hand drive   Black        2  
3         Automatic          4x4  04-May        Left wheel   White        0  
4         Automatic        Front  04-May        Left wheel  Silver        4  
...             ...          ...     ...               ...     ...      ...  
19232        Manual         Rear  02-Mar        Left wheel  Silver        5  
19233     Tiptronic        Front  04-May        Left wheel     Red        8  
19234     Automatic        Front  04-May        Left wheel    Grey        4  
19235     Automatic        Front  04-May        Left wheel   Black        4  
19236     Automatic        Front  04-May        Left wheel   White       12  

[19237 rows x 18 columns]

# Data Cleaning

### Head

In [3]:
data.head()

ID  Price  Levy Manufacturer    Model  Prod. year   Category  \
0  45654403  13328  1399        LEXUS   RX 450        2010       Jeep   
1  44731507  16621  1018    CHEVROLET  Equinox        2011       Jeep   
2  45774419   8467     -        HONDA      FIT        2006  Hatchback   
3  45769185   3607   862         FORD   Escape        2011       Jeep   
4  45809263  11726   446        HONDA      FIT        2014  Hatchback   

  Leather interior Fuel type Engine volume    Mileage  Cylinders  \
0              Yes    Hybrid           3.5  186005 km        6.0   
1               No    Petrol             3  192000 km        6.0   
2               No    Petrol           1.3  200000 km        4.0   
3              Yes    Hybrid           2.5  168966 km        4.0   
4              Yes    Petrol           1.3   91901 km        4.0   

  Gear box type Drive wheels   Doors             Wheel   Color  Airbags  
0     Automatic          4x4  04-May        Left wheel  Silver       12  
1     Tiptronic          4x4  04-May        Left wheel   Black        8  
2      Variator        Front  04-May  Right-hand drive   Black        2  
3     Automatic          4x4  04-May        Left wheel   White        0  
4     Automatic        Front  04-May        Left wheel  Silver        4

## Shape

In [4]:
data.shape

(19237, 18)

We have 19237 rows and 18 columns in our dataset.

## Null Value

In [5]:
data.isnull().sum()

ID                  0
Price               0
Levy                0
Manufacturer        0
Model               0
Prod. year          0
Category            0
Leather interior    0
Fuel type           0
Engine volume       0
Mileage             0
Cylinders           0
Gear box type       0
Drive wheels        0
Doors               0
Wheel               0
Color               0
Airbags             0
dtype: int64

We have no null value in our dataset.

In [6]:
data.describe().T

count          mean            std         min         25%  \
ID          19237.0  4.557654e+07  936591.422799  20746880.0  45698374.0   
Price       19237.0  1.855593e+04  190581.269684         1.0      5331.0   
Prod. year  19237.0  2.010913e+03       5.668673      1939.0      2009.0   
Cylinders   19237.0  4.582991e+00       1.199933         1.0         4.0   
Airbags     19237.0  6.582627e+00       4.320168         0.0         4.0   

                   50%         75%         max  
ID          45772308.0  45802036.0  45816654.0  
Price          13172.0     22075.0  26307500.0  
Prod. year      2012.0      2015.0      2020.0  
Cylinders          4.0         4.0        16.0  
Airbags            6.0        12.0        16.0

## Datatypes

In [7]:
data.dtypes

ID                    int64
Price                 int64
Levy                 object
Manufacturer         object
Model                object
Prod. year            int64
Category             object
Leather interior     object
Fuel type            object
Engine volume        object
Mileage              object
Cylinders           float64
Gear box type        object
Drive wheels         object
Doors                object
Wheel                object
Color                object
Airbags               int64
dtype: object

Here I'll approach with two ways:
1. If Number value with onject datatypes then use `type conversion`.
2. If string value use `Label Encoder`.


### 1. Levy

In [8]:
# Unique Value
len(data.Levy.unique())

559

In [9]:
# Value Count
data.Levy.value_counts()

-       5819
765      486
891      461
639      410
640      405
        ... 
3156       1
2908       1
1279       1
1719       1
1901       1
Name: Levy, Length: 559, dtype: int64

We can see that `-` at  5891 places. So I am going to replace it by mean of other values.

In [10]:
# lets calculate mean
tot = []
for i in data.Levy:
    if i !="-":
        tot.append(int(i))
LevyAvg = sum(tot)/len(tot)

In [11]:
LevyAvg

906.8381278879118

In [12]:
# Now I am replacing "-" by 906.8381
data.Levy = data.Levy.replace("-",LevyAvg)

# convert datatype object to float
data.Levy = data.Levy.astype(float)

In [13]:
data.Levy.dtypes

dtype('float64')

### 2. Manufacturer

In [14]:
# Unique Value
data.Manufacturer.unique()

array(['LEXUS', 'CHEVROLET', 'HONDA', 'FORD', 'HYUNDAI', 'TOYOTA',
       'MERCEDES-BENZ', 'OPEL', 'PORSCHE', 'BMW', 'JEEP', 'VOLKSWAGEN',
       'AUDI', 'RENAULT', 'NISSAN', 'SUBARU', 'DAEWOO', 'KIA',
       'MITSUBISHI', 'SSANGYONG', 'MAZDA', 'GMC', 'FIAT', 'INFINITI',
       'ALFA ROMEO', 'SUZUKI', 'ACURA', 'LINCOLN', 'VAZ', 'GAZ',
       'CITROEN', 'LAND ROVER', 'MINI', 'DODGE', 'CHRYSLER', 'JAGUAR',
       'ISUZU', 'SKODA', 'DAIHATSU', 'BUICK', 'TESLA', 'CADILLAC',
       'PEUGEOT', 'BENTLEY', 'VOLVO', 'სხვა', 'HAVAL', 'HUMMER', 'SCION',
       'UAZ', 'MERCURY', 'ZAZ', 'ROVER', 'SEAT', 'LANCIA', 'MOSKVICH',
       'MASERATI', 'FERRARI', 'SAAB', 'LAMBORGHINI', 'ROLLS-ROYCE',
       'PONTIAC', 'SATURN', 'ASTON MARTIN', 'GREATWALL'], dtype=object)

In [15]:
# Value Count
data.Manufacturer.value_counts()

HYUNDAI          3769
TOYOTA           3662
MERCEDES-BENZ    2076
FORD             1111
CHEVROLET        1069
                 ... 
TESLA               1
PONTIAC             1
SATURN              1
ASTON MARTIN        1
GREATWALL           1
Name: Manufacturer, Length: 65, dtype: int64

In [16]:
leManu = LabelEncoder()
leManu.fit(data.Manufacturer)
data.Manufacturer = leManu.transform(data.Manufacturer)

In [17]:
data.Manufacturer.dtypes

dtype('int32')

In [18]:
data.Manufacturer

0        32
1         8
2        21
3        16
4        21
         ..
19232    36
19233    23
19234    23
19235     8
19236    23
Name: Manufacturer, Length: 19237, dtype: int32

### 3. Model

In [19]:
data.Model.unique()

array(['RX 450', 'Equinox', 'FIT', ..., 'E 230 124', 'RX 450 F SPORT',
       'Prius C aqua'], dtype=object)

In [20]:
data.Model.value_counts()

Prius                    1083
Sonata                   1079
Camry                     938
Elantra                   922
E 350                     542
                         ... 
Feroza                      1
C-MAX C-MAX                 1
X1 4X4                      1
Land Cruiser Prado RX       1
Prius C aqua                1
Name: Model, Length: 1590, dtype: int64

In [21]:
leModel = LabelEncoder()
leModel.fit(data.Model)
data.Model = leModel.transform(data.Model)

In [22]:
data.Model.dtypes

dtype('int32')

In [23]:
data.Model

0        1242
1         658
2         684
3         661
4         684
         ... 
19232     385
19233    1334
19234    1442
19235     456
19236    1334
Name: Model, Length: 19237, dtype: int32

### 3. Leather Ineterior

In [24]:
# Unique Value
data["Leather interior"].unique()

array(['Yes', 'No'], dtype=object)

In [25]:
# Value Count
data["Leather interior"].value_counts()

Yes    13954
No      5283
Name: Leather interior, dtype: int64

Here we have value only **No** and **No**
I'm going map:
* `Yes` = 1
* `No` = 0

In [26]:
data["Leather interior"] = data["Leather interior"].map({"Yes":1,"No":0})
data["Leather interior"]

0        1
1        0
2        0
3        1
4        1
        ..
19232    1
19233    1
19234    1
19235    1
19236    1
Name: Leather interior, Length: 19237, dtype: int64

In [27]:
data["Leather interior"].dtype

dtype('int64')

### 4. Mileage

In [28]:
data.Mileage

0        186005 km
1        192000 km
2        200000 km
3        168966 km
4         91901 km
           ...    
19232    300000 km
19233    161600 km
19234    116365 km
19235     51258 km
19236    186923 km
Name: Mileage, Length: 19237, dtype: object

Here all values are ends with ` Km` , so here I'll remove ` km` with help of string spliting. each value, and convert datatype to float 

In [29]:
# Demo 
str = "12000 Km"
str = str[:-3] # String spliting
str = float(str) # Type conversion
print(str)
print(type(str))

12000.0
<class 'float'>


In [30]:
Mile = []
for s in data.Mileage:
    s = s[:-3]
    s = float(s)
    Mile.append(s)
data.Mileage = Mile  

In [31]:
data.Mileage

0        186005.0
1        192000.0
2        200000.0
3        168966.0
4         91901.0
           ...   
19232    300000.0
19233    161600.0
19234    116365.0
19235     51258.0
19236    186923.0
Name: Mileage, Length: 19237, dtype: float64

### 5. Engine volume

In [32]:
print(list(data["Engine volume"][:50]))

['3.5', '3', '1.3', '2.5', '1.3', '2', '1.8', '2.4', '2.5', '3.5', '3.5', '4', '1.6', '3.5', '2', '2', '1.8', '1.8', '3.3', '3.5', '1.6', '1.8', '2.5', '2.0 Turbo', '1.8', '2.2 Turbo', '4.7', '1.5', '1.6', '3.3', '2.0 Turbo', '3', '4.4', '3.3', '3.0 Turbo', '2', '2', '1.3', '1.6', '3.5', '2.4', '2.5', '1.4 Turbo', '2.4', '3.6', '4', '1.5', '2', '1.5', '2.3']


In Engine volume almost all value are in numeric but there it problem. That is in some cases ` Turbo` suffix added.
I'm going to handle this:
* Case 1: Value with No `Turbo` suffix I will convert datatypes.
* Case 2: Value with Turbe suffix I will remove `Turbo`, then convert datatypes.


In [33]:
Eng = []
for s in data["Engine volume"]:
    if "Turbo" in s:
        string = s[:3]
        Eng.append(float(string))
    else:
        Eng.append(float(s))
        
data["Engine volume"] = Eng

In [34]:
data["Engine volume"].dtypes

dtype('float64')

### "Fuel type","Gear box type","Drive wheels","Doors","Wheel","Color"

In [35]:
# Label Encoding using for loop
objDtypes = ["Fuel type","Gear box type","Drive wheels","Doors","Wheel","Color","Category"]

In [36]:
uni = {}
for i in objDtypes:
    val = data[i].unique()
    uni[i] = val
uni

{'Fuel type': array(['Hybrid', 'Petrol', 'Diesel', 'CNG', 'Plug-in Hybrid', 'LPG',
        'Hydrogen'], dtype=object),
 'Gear box type': array(['Automatic', 'Tiptronic', 'Variator', 'Manual'], dtype=object),
 'Drive wheels': array(['4x4', 'Front', 'Rear'], dtype=object),
 'Doors': array(['04-May', '02-Mar', '>5'], dtype=object),
 'Wheel': array(['Left wheel', 'Right-hand drive'], dtype=object),
 'Color': array(['Silver', 'Black', 'White', 'Grey', 'Blue', 'Green', 'Red',
        'Sky blue', 'Orange', 'Yellow', 'Brown', 'Golden', 'Beige',
        'Carnelian red', 'Purple', 'Pink'], dtype=object),
 'Category': array(['Jeep', 'Hatchback', 'Sedan', 'Microbus', 'Goods wagon',
        'Universal', 'Coupe', 'Minivan', 'Cabriolet', 'Limousine',
        'Pickup'], dtype=object)}

So **"Fuel type","Gear box type","Drive wheels","Doors","Wheel","Color","Category"** these are columns with `string value`, I will apply `LabelEncoding` with for loop

In [37]:
for i in objDtypes:
    le = LabelEncoder()
    le.fit(data[i])
    data[i] = le.transform(data[i])

In [38]:
data.dtypes

ID                    int64
Price                 int64
Levy                float64
Manufacturer          int32
Model                 int32
Prod. year            int64
Category              int32
Leather interior      int64
Fuel type             int32
Engine volume       float64
Mileage             float64
Cylinders           float64
Gear box type         int32
Drive wheels          int32
Doors                 int32
Wheel                 int32
Color                 int32
Airbags               int64
dtype: object

In [39]:
data.head()

ID  Price         Levy  Manufacturer  Model  Prod. year  Category  \
0  45654403  13328  1399.000000            32   1242        2010         4   
1  44731507  16621  1018.000000             8    658        2011         4   
2  45774419   8467   906.838128            21    684        2006         3   
3  45769185   3607   862.000000            16    661        2011         4   
4  45809263  11726   446.000000            21    684        2014         3   

   Leather interior  Fuel type  Engine volume   Mileage  Cylinders  \
0                 1          2            3.5  186005.0        6.0   
1                 0          5            3.0  192000.0        6.0   
2                 0          5            1.3  200000.0        4.0   
3                 1          2            2.5  168966.0        4.0   
4                 1          5            1.3   91901.0        4.0   

   Gear box type  Drive wheels  Doors  Wheel  Color  Airbags  
0              0             0      1      0     12       12  
1              2             0      1      0      1        8  
2              3             1      1      1      1        2  
3              0             0      1      0     14        0  
4              0             1      1      0     12        4

I have converted all data to numveric type.

In [40]:
## Drop Id columns
data = data.drop(["ID"],axis=1)
data.head()

Price         Levy  Manufacturer  Model  Prod. year  Category  \
0  13328  1399.000000            32   1242        2010         4   
1  16621  1018.000000             8    658        2011         4   
2   8467   906.838128            21    684        2006         3   
3   3607   862.000000            16    661        2011         4   
4  11726   446.000000            21    684        2014         3   

   Leather interior  Fuel type  Engine volume   Mileage  Cylinders  \
0                 1          2            3.5  186005.0        6.0   
1                 0          5            3.0  192000.0        6.0   
2                 0          5            1.3  200000.0        4.0   
3                 1          2            2.5  168966.0        4.0   
4                 1          5            1.3   91901.0        4.0   

   Gear box type  Drive wheels  Doors  Wheel  Color  Airbags  
0              0             0      1      0     12       12  
1              2             0      1      0      1        8  
2              3             1      1      1      1        2  
3              0             0      1      0     14        0  
4              0             1      1      0     12        4

### Describe

In [41]:
data.describe().T

count          mean           std     min      25%  \
Price             19237.0  1.855593e+04  1.905813e+05     1.0   5331.0   
Levy              19237.0  9.068381e+02  3.857335e+02    87.0    730.0   
Manufacturer      19237.0  3.308068e+01  1.776646e+01     0.0     21.0   
Model             19237.0  8.625923e+02  4.108123e+02     0.0    537.0   
Prod. year        19237.0  2.010913e+03  5.668673e+00  1939.0   2009.0   
Category          19237.0  6.265582e+00  2.790570e+00     0.0      4.0   
Leather interior  19237.0  7.253730e-01  4.463377e-01     0.0      0.0   
Fuel type         19237.0  3.432396e+00  1.807388e+00     0.0      2.0   
Engine volume     19237.0  2.307990e+00  8.778045e-01     0.0      1.8   
Mileage           19237.0  1.532236e+06  4.840387e+07     0.0  70139.0   
Cylinders         19237.0  4.582991e+00  1.199933e+00     1.0      4.0   
Gear box type     19237.0  5.363102e-01  8.966066e-01     0.0      0.0   
Drive wheels      19237.0  9.088735e-01  5.678746e-01     0.0      1.0   
Doors             19237.0  9.662629e-01  2.142637e-01     0.0      1.0   
Wheel             19237.0  7.714301e-02  2.668252e-01     0.0      0.0   
Color             19237.0  7.776005e+00  5.363884e+00     0.0      1.0   
Airbags           19237.0  6.582627e+00  4.320168e+00     0.0      4.0   

                            50%       75%           max  
Price              13172.000000   22075.0  2.630750e+07  
Levy                 906.838128     917.0  1.171400e+04  
Manufacturer          32.000000      54.0  6.400000e+01  
Model                834.000000    1226.0  1.589000e+03  
Prod. year          2012.000000    2015.0  2.020000e+03  
Category               7.000000       9.0  1.000000e+01  
Leather interior       1.000000       1.0  1.000000e+00  
Fuel type              5.000000       5.0  6.000000e+00  
Engine volume          2.000000       2.5  2.000000e+01  
Mileage           126000.000000  188888.0  2.147484e+09  
Cylinders              4.000000       4.0  1.600000e+01  
Gear box type          0.000000       1.0  3.000000e+00  
Drive wheels           1.000000       1.0  2.000000e+00  
Doors                  1.000000       1.0  2.000000e+00  
Wheel                  0.000000       0.0  1.000000e+00  
Color                  7.000000      12.0  1.500000e+01  
Airbags                6.000000      12.0  1.600000e+01

# Spliting Data Into featues and labels
* **Features**: Levy,Manufacturer,Model,Prod. year,Category,Leather interior,Fuel type,Engine volume,Mileage,Cylinders,Gear box type,Drive wheels,Doors,Wheel,Color,Airbags.
* **Labels**: Price

In [42]:
x = data.drop(["Price"],axis=1)
y = data["Price"]

In [43]:
x.head()

Levy  Manufacturer  Model  Prod. year  Category  Leather interior  \
0  1399.000000            32   1242        2010         4                 1   
1  1018.000000             8    658        2011         4                 0   
2   906.838128            21    684        2006         3                 0   
3   862.000000            16    661        2011         4                 1   
4   446.000000            21    684        2014         3                 1   

   Fuel type  Engine volume   Mileage  Cylinders  Gear box type  Drive wheels  \
0          2            3.5  186005.0        6.0              0             0   
1          5            3.0  192000.0        6.0              2             0   
2          5            1.3  200000.0        4.0              3             1   
3          2            2.5  168966.0        4.0              0             0   
4          5            1.3   91901.0        4.0              0             1   

   Doors  Wheel  Color  Airbags  
0      1      0     12       12  
1      1      0      1        8  
2      1      1      1        2  
3      1      0     14        0  
4      1      0     12        4

In [44]:
y.head()

0    13328
1    16621
2     8467
3     3607
4    11726
Name: Price, dtype: int64

## Spliting data into train and Test set

In [45]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [46]:
len(x_train),len(x_test),len(y_train),len(y_test)

(15389, 3848, 15389, 3848)

We are giving 15389 data to train model, and going to evaluate on 3848 data

## Feature Scaling

In [47]:
sc  = StandardScaler()
x_train = pd.DataFrame(sc.fit_transform(x_train))
x_test = pd.DataFrame(sc.transform(x_test))

In [48]:
x_train

0         1         2         3         4         5         6   \
0     -0.342678  1.405025 -1.039444  0.364707  0.978136  0.612072 -0.785166   
1      1.022924 -1.411988 -0.988375 -0.699707 -0.814985  0.612072 -1.337428   
2      1.340568  0.165539 -0.635759 -0.167500  0.978136  0.612072 -1.337428   
3     -0.714621 -1.411988  0.553408  0.187305 -0.814985  0.612072 -1.337428   
4      0.004392 -1.581009  1.630711 -1.941523 -0.814985  0.612072  0.319356   
...         ...       ...       ...       ...       ...       ...       ...   
15384 -0.068471  1.179664 -1.494197  1.251718 -0.814985  0.612072  0.871618   
15385 -0.804213  1.405025 -1.039444  1.251718  0.978136  0.612072 -0.785166   
15386  0.026551 -1.581009  1.630711  0.542109 -0.814985  0.612072 -1.337428   
15387  0.004392 -1.581009 -1.885723 -1.054511  0.978136  0.612072  0.871618   
15388  0.004392  0.165539 -1.309378 -2.828534  0.978136 -1.633796  0.871618   

             7         8         9         10        11        12        13  \
0      0.218100 -0.027271 -0.485097 -0.597371  0.150973  0.150671 -0.286852   
1     -0.350077 -0.030280 -0.485097 -0.597371  0.150973  0.150671 -0.286852   
2      1.354454 -0.027996  1.188704 -0.597371 -1.614469  0.150671 -0.286852   
3     -0.350077 -0.027412 -0.485097 -0.597371  0.150973  0.150671 -0.286852   
4      0.786277 -0.031312  1.188704  1.633925 -1.614469  0.150671 -0.286852   
...         ...       ...       ...       ...       ...       ...       ...   
15384 -0.804619 -0.031575 -0.485097 -0.597371  0.150973  0.150671 -0.286852   
15385  0.218100 -0.031109 -0.485097 -0.597371  0.150973  0.150671 -0.286852   
15386  0.786277 -0.029585  1.188704 -0.597371 -1.614469  0.150671 -0.286852   
15387  0.218100 -0.028551  1.188704  1.633925  1.916414  0.150671  3.486114   
15388 -0.122807 -0.025568 -0.485097  0.518277  1.916414  0.150671 -0.286852   

             14        15  
0      1.159130  1.254361  
1      0.785523 -0.594687  
2     -1.269313  1.254361  
3     -0.148493 -0.594687  
4      0.785523 -0.132425  
...         ...       ...  
15384  0.598720 -0.594687  
15385  1.159130  1.254361  
15386  1.159130 -1.519210  
15387  1.159130  1.254361  
15388 -1.269313 -0.594687  

[15389 rows x 16 columns]

# Training Model

In [49]:
model = RandomForestRegressor(random_state=42)
model.fit(x_train,y_train)
model.score(x_test,y_test)

-40.66672304914635

In [50]:
ypred = model.predict(x_test)
ypred

array([24718.4 , 11647.14,  5452.78, ...,   314.  , 14506.4 , 14548.75])

####  Evaluate model 

In [51]:
# r2_score
r2_score(y_test,ypred)

-40.66672304914635

# Hyperparameters Tunning

In [52]:
n_estimators = [5,10,20,30,40,50,60,70,80,90,100]
for i in n_estimators:
    mdl = RandomForestRegressor(n_estimators=i)
    mdl.fit(x_train,y_train)
    s = mdl.score(x_test,y_test)
    print(f"The score of model is {s} with n_estimators {i} ")

The score of model is -45.51915441606863 with n_estimators 5 
The score of model is -22.446789176327055 with n_estimators 10 
The score of model is -55.901900426393155 with n_estimators 20 
The score of model is -62.27838281709403 with n_estimators 30 
The score of model is -14.547236286240263 with n_estimators 40 
The score of model is -46.02025469653213 with n_estimators 50 
The score of model is -28.70197325118746 with n_estimators 60 
The score of model is -47.35458939223819 with n_estimators 70 
The score of model is -39.4984567111281 with n_estimators 80 
The score of model is -55.466874954358765 with n_estimators 90 
The score of model is -34.854815498196444 with n_estimators 100 


we have got best score for  `n_estimators=40`

In [53]:
max_depth = [5,7,10,13,15,17,20,22,25,28,30,40,50]
for i in max_depth:
    mdl = RandomForestRegressor(n_estimators=40,max_depth=i)
    mdl.fit(x_train,y_train)
    s = mdl.score(x_test,y_test)
    print(f"The score of model is {s} with max_depth {i} ")

The score of model is -37.025687683264906 with max_depth 5 
The score of model is -59.175252601043645 with max_depth 7 
The score of model is -50.720553843127966 with max_depth 10 
The score of model is -50.24543257425519 with max_depth 13 
The score of model is -32.555904681536994 with max_depth 15 
The score of model is -26.131729911366264 with max_depth 17 
The score of model is -29.749662599914213 with max_depth 20 
The score of model is -93.94035954008953 with max_depth 22 
The score of model is -82.38076804387703 with max_depth 25 
The score of model is -39.13531916314249 with max_depth 28 
The score of model is -85.20011923643989 with max_depth 30 
The score of model is -58.94812527527128 with max_depth 40 
The score of model is -42.22169218024597 with max_depth 50 


We have got best score for  `max_depth=17`

In [55]:
max_feature = [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0]
for i in max_feature:
    mdl = RandomForestRegressor(n_estimators=40,max_depth=17,max_features=i)
    mdl.fit(x_train,y_train)
    s = mdl.score(x_test,y_test)
    print(f"The score of model is {s} with max feature {i}. ")

The score of model is -4.415848549112867 with max feature 0.1. 
The score of model is -8.751005024744742 with max feature 0.2. 
The score of model is -6.550410115830636 with max feature 0.3. 
The score of model is -13.0242987731656 with max feature 0.4. 
The score of model is -7.230678010727013 with max feature 0.5. 
The score of model is -9.077466397174275 with max feature 0.6. 
The score of model is -13.52471598952811 with max feature 0.7. 
The score of model is -20.350246814770784 with max feature 0.8. 
The score of model is -65.75803927601396 with max feature 0.9. 
The score of model is -116.86267509404281 with max feature 1.0. 


We have got best score for  `max_depth=0.1`

In [56]:
min_samples_split = [0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0]
for i in min_samples_split:
    mdl = RandomForestRegressor(n_estimators=40,max_depth=17,max_features=0.1,min_samples_split=i)
    mdl.fit(x_train,y_train)
    s = mdl.score(x_test,y_test)
    print(f"The score of model is {s} with min feature {i}. ")

The score of model is 0.12135839026223727 with min feature 0.1. 
The score of model is 0.07754326366533748 with min feature 0.2. 
The score of model is 0.08343388383701533 with min feature 0.3. 
The score of model is 0.057263183649814864 with min feature 0.4. 
The score of model is 0.051211494409765246 with min feature 0.5. 
The score of model is 0.0032440834231227678 with min feature 0.6. 
The score of model is -0.008379728090617267 with min feature 0.7. 
The score of model is -0.016682297330984808 with min feature 0.8. 
The score of model is -0.016886795109537722 with min feature 0.9. 
The score of model is -0.009703403823108303 with min feature 1.0. 


We have got best score for `min_samples_split=0.7`

In [57]:
# Model with best hyperparameters
best_model = RandomForestRegressor(n_estimators=40,max_depth=17,max_features=0.1,min_samples_split=0.7,random_state=42)
best_model.fit(x_train,y_train)
best_model.score(x_test,y_test)

-0.00837507862086384

In [58]:
ypredBest = best_model.predict(x_test)
ypredBest

array([18604.69174248, 18604.69174248, 18604.69174248, ...,
       18604.69174248, 18604.69174248, 18604.69174248])

## Evaluation of best model

In [59]:
# r2_score
r2_score(y_test,ypredBest)

-0.00837507862086384

In [60]:
# mean_absolute_error
mean_absolute_error(y_test,ypredBest)

12405.453764390833

In [61]:
# mean_squared_error
mean_squared_error(y_test,ypredBest)

314206260.8281274

# Conclusion

We have build Random Forest Regressor model to predict car price.

### Thank you reading....